module/zfs/zfs_crrd.c

```diff
@@ -0,0 +1,230 @@
+}
+
+/* Add value to database. */
+void
+rrd_add(rrd_t *rrd, hrtime_t time, uint64_t txg)
+{
+
+	rrd->rrd_entries[rrd->rrd_tail].rrdd_time = time;
+	rrd->rrd_entries[rrd->rrd_tail].rrdd_txg = txg;
+
+	rrd->rrd_tail = (rrd->rrd_tail + 1) % RRD_MAX_ENTRIES;
+
+	if (rrd->rrd_length < RRD_MAX_ENTRIES) {
+		rrd->rrd_length++;
+	} else {
+		rrd->rrd_head = (rrd->rrd_head + 1) % RRD_MAX_ENTRIES;
+	}
+}
+
+void
+dbrrd_add(dbrrd_t *db, hrtime_t time, uint64_t txg)
+{
+	hrtime_t daydiff, monthdiff;
+
+	daydiff = time - rrd_tail(&db->dbr_days);
+	monthdiff = time - rrd_tail(&db->dbr_months);
+
+	rrd_add(&db->dbr_minutes, time, txg);
+	if (daydiff >= 0 && daydiff >= SEC2NSEC(60 * 3600))
+		rrd_add(&db->dbr_days, time, txg);
+	if (monthdiff >= 0 && monthdiff >= SEC2NSEC(60 * 3600 * 30))
+		rrd_add(&db->dbr_months, time, txg);
+}
+
+/*
+ * XXXosho: We might want to do a binary search here,
+ *          although the data is small, and the routine
+ *          isn't used so often that we stick to simple methods.
+ */
+static const rrd_data_t *
+rrd_query(rrd_t *rrd, hrtime_t tv, dbrrd_rounding_t rounding)
+{
+	const rrd_data_t *data;
+
+	data = NULL;
+	for (size_t i = 0; i < rrd_len(rrd); i++) {
+		const rrd_data_t *cur = rrd_entry(rrd, i);
+
+		if (rounding == DBRRD_FLOOR) {
+			if (tv < cur->rrdd_time) {
+				break;
+			}
+			data = cur;
+		} else {
+			/* DBRRD_CEILING */
+			if (tv <= cur->rrdd_time) {
+				data = cur;
+				break;
+			}
+		}
+	}
+
+	return (data);
+}
+
+static const rrd_data_t *
+dbrrd_min(const rrd_data_t *r1, const rrd_data_t *r2)
+{
+	if (r1 == NULL)
+		return (r2);
+	if (r2 == NULL)
+		return (r1);
+
+	return (r1->rrdd_txg < r2->rrdd_txg ? r1 : r2);
+}
+
+static const rrd_data_t *
+dbrrd_max(const rrd_data_t *r1, const rrd_data_t *r2)
+{
+	if (r1 == NULL)
+		return (r2);
+	if (r2 == NULL)
+		return (r1);
+
+	return (r1->rrdd_txg > r2->rrdd_txg ? r1 : r2);
+}
+
+uint64_t
+dbrrd_query(dbrrd_t *r, hrtime_t tv, dbrrd_rounding_t rounding)
+{
+	const rrd_data_t *data, *dm, *dd, *dy;
+
+	data = NULL;
+	dm = rrd_query(&r->dbr_minutes, tv, rounding);
+	dd = rrd_query(&r->dbr_days, tv, rounding);
+	dy = rrd_query(&r->dbr_months, tv, rounding);
+
+	if (rounding == DBRRD_FLOOR) {
+		data = dbrrd_max(dbrrd_min(dd, dm), dy);
+	} else {
+		data = dbrrd_min(dbrrd_min(dd, dm), dy);
+	}
+
+	return (data == NULL ? 0 : data->rrdd_txg);
+}
+
+void
+rrd_pack(rrd_t *rrd, uint8_t *buffer)
+{
+	rrd_t *ptr = (rrd_t *)buffer;
+
+	memset(buffer, 0, sizeof (*ptr));
+
+	ptr->rrd_head = htonll(rrd->rrd_head);
+	ptr->rrd_tail = htonll(rrd->rrd_tail);
+	ptr->rrd_length = htonll(rrd->rrd_length);
+
+	for (size_t i = 0; i < RRD_MAX_ENTRIES; i++) {
+		ptr->rrd_entries[i].rrdd_time =
+		    htonll(rrd->rrd_entries[i].rrdd_time);
+		ptr->rrd_entries[i].rrdd_txg =
+		    htonll(rrd->rrd_entries[i].rrdd_txg);
+	}
+}
+
+void
+rrd_ntoh(rrd_t *rrd)
+{
+	rrd->rrd_head = ntohll(rrd->rrd_head);
+	rrd->rrd_tail = ntohll(rrd->rrd_tail);
+	rrd->rrd_length = ntohll(rrd->rrd_length);
```

Comment:
While not incorrect, I can't recall other place where ZFS would store things in network order.  DDT does it per-record, for example see `ddt_zap_decompress()`.  I don't insist, just propose to think wider.

Also it seems like this will be a single > 4KB ZAP entry.  I hope MOS ZAP has sufficiently big block size to store it efficiently.